File: java/src/net/omniscimus/profielwerkstuk/mysql/MySQLManager.java
package net.omniscimus.profielwerkstuk.mysql;

import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.SQLException;
import java.util.logging.Level;
import java.util.logging.Logger;
import net.omniscimus.profielwerkstuk.configuration.ConfigValueCache;

/**
 * Verzorgt de verbinding met de MySQL database.
 *
 * @author omniscimus
 */
public class MySQLManager {

    private final SchoolSQL schoolSQL;
    private final ScheduleChangesSQL scheduleChangesSQL;
    private final DatabaseLink databaseLink;

    /**
     * Maakt een nieuwe MySQLManager; opent de verbinding met de MySQL server.
     *
     * @throws SQLException als er geen toegang tot de database verkregen kon
     * worden
     * @throws ClassNotFoundException als het stuurprogramma voor de MySQL
     * server niet gevonden kon worden
     */
    public MySQLManager() throws SQLException, ClassNotFoundException {
	schoolSQL = new SchoolSQL(this);
	scheduleChangesSQL = new ScheduleChangesSQL(this);
	databaseLink = new DatabaseLink(scheduleChangesSQL, schoolSQL);

	openConnection();
    }

    /**
     * Geeft de DatabaseLink die dit programma gebruikt.
     *
     * @return de te gebruiken DatabaseLink
     */
    public DatabaseLink getDatabaseLink() {
	return databaseLink;
    }

    /**
     * Geeft de SchoolSQL die dit programma gebruikt.
     *
     * @return de te gebruiken SchoolSQL
     */
    public SchoolSQL getSchoolSQL() {
	return schoolSQL;
    }

    /**
     * Geeft de ScheduleChangesSQL die dit programma gebruikt.
     *
     * @return de te gebruiken ScheduleChangesSQL
     */
    public ScheduleChangesSQL getScheduleChangesSQL() {
	return scheduleChangesSQL;
    }

    private Connection connection;

    /**
     * Opent de verbinding met de MySQL server. Gebruikt de gegevens die in
     * ConfigValueCache staan.
     *
     * @return de gemaakte verbinding
     * @throws SQLException als er geen verbinding kon worden geopend
     * @throws ClassNotFoundException als het stuurprogramma voor de MySQL
     * server niet gevonden kon worden
     */
    private Connection openConnection() throws SQLException, ClassNotFoundException {
	Class.forName("com.mysql.jdbc.Driver");
	connection = DriverManager.getConnection("jdbc:mysql://" + ConfigValueCache.getSQLHostname() + ":" + ConfigValueCache.getSQLPort(),
		ConfigValueCache.getSQLUsername(), ConfigValueCache.getSQLPassword());
	return connection;
    }

    /**
     * Geeft de huidige MySQL verbinding, of maakt een nieuwe als er nog geen is
     * of de huidige niet bruikbaar is.
     *
     * @return een bruikbare verbinding met de MySQL database
     * @throws SQLException als er geen verbinding kon worden geopend
     * @throws ClassNotFoundException als het stuurprogramma voor de MySQL
     * server niet gevonden kon worden
     */
    public Connection getConnection() throws SQLException, ClassNotFoundException {
	if (connection != null && !connection.isClosed() && connection.isValid(3)) {
	    return connection;
	} else {
	    return openConnection();
	}
    }

    /**
     * Sluit de verbinding met de MySQL-database.
     */
    public void closeConnection() {
	try {
	    if (connection != null && !connection.isClosed() && connection.isValid(3)) {
		connection.close();
	    }
	} catch (Exception ex) {
	    Logger.getLogger(MySQLManager.class.getName()).log(Level.SEVERE, null, ex);
	}
    }

}

File: php/sql/MySQL_Manager.php
<?php
namespace ScheduleChanges;

require_once 'School_SQL.php';

/**
 * Manager voor de verbinding met de MySQL database.
 *
 * @author omniscimus
 */
class MySQL_Manager {

    private $connection;
    private $school_SQL;

    /**
     * Maakt een nieuwe MySQL_Manager.
     */
    function __construct() {
        $this->school_SQL = new School_SQL($this);
    }
    
    function getSchoolSQL() {
        return $this->school_SQL;
    }

    /**
     * Geeft de MySQL-verbinding.
     * 
     * @return mysqli de verbinding die gebruikt kan worden voor queries
     */
    function getConnection() {
        if (!isset($this->connection)) {
            $this->connect();
        }
        return $this->connection;
    }
    
    /**
     * Sluit de MySQL verbinding (als die bestaat).
     */
    function closeConnection() {
        if(isset($this->connection)) {
            $this->connection->close();
        }
    }

    /**
     * Verbindt met de MySQL server.
     * 
     * @throws Exception als het verbinden is mislukt
     */
    function connect() {
        $config = include 'config.php';
        $connection = new mysqli(
                $config["mysql_hostname"], $config["mysql_username"], $config["mysql_password"], NULL, $config["mysql_port"]);
        if ($connection->connect_error) {
            throw new Exception("Verbinden met de database is mislukt.");
        } else {
            $this->connection = $connection;
        }
    }

}
